### PR TITLE
ref(StructuredEventData): Rewrite expand/collapse state into a provider

### PR DESCRIPTION
**Pull request by ryan953**
Pushed the expand/collapse state into a provider, which accepts a callback so external systems can save the state and pass it back in on mount so that the component will render fresh with the state from an old component.

There's a `useRef` to track the overall state of things in the provider, and each `<CollapsibleValue>` has it's own `setState` so when something is toggled we don't endup having react look at the whole tree.

### CODE REVIEW

File: static/app/components/structuredEventData/index.tsx [hunk 1]
```diff
@@ -29,73 +75,85 @@ interface StructuredDataProps {
...
+  withOnlyFormattedText = false,
 }: StructuredDataProps) {
+  const getInitialExpandedPaths = useCallback(() => {
+    return (
+      initialExpandedPaths ??
+      (forceDefaultExpand === false ||
+      (forceDefaultExpand === undefined && !maxDefaultDepth)
+        ? []
+        : getDefaultExpanded(Math.max(1, maxDefaultDepth), value))
+    );
```
Comment (ryan953) [inline static/app/components/structuredEventData/index.tsx:95]: I didn't write unit tests for this, but the "Auto-Expanded items" story renders the same as before. See https://github.com/getsentry/sentry/pull/74481 as a reference.

File: static/app/components/structuredEventData/collapsibleValue.tsx [hunk 2]
```diff
@@ -1,44 +1,35 @@
...
+  path,
   prefix = null,
-  depth,
-  maxDefaultDepth,
-  forceDefaultExpand,
-}: CollapsibleValueProps) {
+}: Props) {
+  const {collapse, expand, isExpanded: isInitiallyExpanded} = useExpandedState({path});
+  const [isExpanded, setIsExpanded] = useState(isInitiallyExpanded);
+  // const isExpanded = expandedPaths.includes(path);
```
Comment (malwilley) [inline static/app/components/structuredEventData/collapsibleValue.tsx:27]: Left a little something here

File: static/app/utils/replays/countDomNodes.tsx [hunk 3]
```diff
@@ -26,6 +27,7 @@ export default function countDomNodes({
   const frameStep = Math.max(Math.round(length * 0.007), 1);
 
   let prevIds: number[] = [];
+  useOrganization();
```
Comment (malwilley) [inline static/app/utils/replays/countDomNodes.tsx:30]: And this seems unrelated

File: static/app/components/structuredEventData/useExpandedState.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import type {ReactNode} from 'react';
+import {createContext, useCallback, useContext, useMemo, useRef} from 'react';
+
+import {uniq} from 'sentry/utils/array/uniq';
+
+type State = 'expanded' | 'collapsed';
+
+const context = createContext<{
```
Comment (malwilley) [inline static/app/components/structuredEventData/useExpandedState.tsx:8]: nit: casing would usually be `Context`

File: static/app/components/structuredEventData/index.tsx [hunk 5]
```diff
@@ -15,12 +18,55 @@ export type StructedEventDataConfig = {
   renderString?: (value: string) => string;
 };
 
-interface StructuredDataProps {
-  maxDefaultDepth: number;
-  withAnnotatedText: boolean;
+interface BaseProps {
```
Comment (malwilley) [inline static/app/components/structuredEventData/index.tsx:21]: Thanks for the descriptions! 👏

File: static/app/components/structuredEventData/useExpandedState.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  initialExpandedPaths: () => string[];
+  onToggleExpand?: (expandedPaths: string[], path: string, state: State) => void;
+}
+
+export function ExpandedStateContextProvider({
+  children,
+  initialExpandedPaths,
+  onToggleExpand,
+}: Props) {
+  const expandedRef = useRef<string[]>(initialExpandedPaths());
```
Comment (malwilley) [inline static/app/components/structuredEventData/useExpandedState.tsx:29]: Is there a reason we need to keep this `expandedRef` up to date when expanding/collapsing? It seems a bit confusing to me if we have two sources of truth in the component. Could we just use it for the initial state and leave it be after that?

File: static/app/components/structuredEventData/index.tsx (hunk 1 above)
Comment (malwilley) [reply static/app/components/structuredEventData/index.tsx:95]: Should we write a test for this? Seems like a lot of extra logic that could regress

Comment (ryan953) [reply static/app/components/structuredEventData/index.tsx:95]: tested!

File: static/app/components/structuredEventData/useExpandedState.tsx (hunk 6 above)
Comment (ryan953) [reply static/app/components/structuredEventData/useExpandedState.tsx:29]: yes! it makes the callback api so much easier.

The main use-case for the callback is so we can keep track of everything that's expanded, and then re-render with that same state in the future. This is especially needed in breadcrumbs inside the replay details page where we use a virtual list, as the user scrolls list items are being unmounted and re-mounted all the time.

Without the callback a callsite would need to do the `.concat` and `.filter` calls (just a few lines below here). But with the ref the callback takes that on instead. Then we can shove the whole state into another ref that contains a list of states for our virtual list to use.

Comment (malwilley) [reply static/app/components/structuredEventData/useExpandedState.tsx:29]: Ahh okay, I see the utility now 👍

**Review by malwilley — APPROVED** [review]
